src/CostFunction3D.cpp

```diff
@@ -2614,57 +2614,62 @@ void CostFunction3D::calcHmatrix()
 	  					cIndex = INDEX(iNode, jNode, kNode, iDim, jDim, varDim, var);
               weight = ibasis * jbasis * kbasis * obsVector[wgt_index];
               H[hi] = weight;
-              JH[hi] = cIndex;
-              mIncr[cIndex]+=1;
-              // if(cIndex==0) {cout << "cindex==0 m: " << m << "\n";}
-              mTmp[hi]=m;
+              JH[hi] = (uint32_t) cIndex;
               hi++;
             }
           }
         }
       }
     }
   }
-  I2H = new integer [nonzeros];
-
-  //for (n=0;n<=8;n++) {cout << " mIncr: " << mIncr[n] << " \n"; }
-  mPtr[0]=0;
-  for (n=1;n<=nState;n++){
-     mPtr[n] = mPtr[n-1]+mIncr[n-1];
-  }
-  //for (n=0;n<=12;n++) {cout << " mPtr: " << mPtr[n] << " \n"; }
-  //cout << "mPtr[nState]: " << mPtr[nState] << " \n";
-  for (n=0;n<nState;n++){mIncr[n]=0;}
-  for (hi=0;hi<nonzeros;hi++){
-    cIndex = JH[hi];
-    dst = mPtr[cIndex]+mIncr[cIndex];
-    I2H[dst] = hi;
-    mVal[dst] = mTmp[hi];
-    mIncr[cIndex]+=1;
+  std::cout << "CostFunction3D::calcHmatrix: after big loop" << std::endl;
+  std::cout << "CostFunction3D::calcHmatrix: Before construction of H^t" << std::endl;
+  /* Calculate the indicies for the H^t matrix */
+  for (uint64_t n=0;n<nState;n++) {IHt[n]=0;}
+  for (uint64_t hi=0;hi<nnz;hi++) {IHt[JH[hi]]++;}
+  std::cout << "CostFunction3D::calcHmatrix: After counting rows H^t" << std::endl;
+  uint64_t cusum=0;
+  for (uint64_t col=0; col < nState;col++) {
+     uint64_t temp = IHt[col];
+     IHt[col] = cusum;
+     cusum += temp;
+  } 
+  std::cout << "CostFunction3D::calcHmatrix: before H^t assignment" << std::endl;
+  IHt[nState]=cusum;
+  for (uint64_t row = 0;row<mObs;row++) {
+      for (uint64_t jj = IH[row];jj< IH[row+1];jj++) {
+	  uint64_t col = (uint64_t) JH[jj];
+          uint64_t dest = IHt[col];
+	  JHt[dest] = (uint32_t) row;
+	  Ht[dest] = H[jj];
+	  IHt[col] += 1;
+      }
   }
```

Comment:
I think here you are constructing the CSR format to save row/column index for the H^t matrix. Could you explain what is the difference between the original `mPtr,mVal,I2H` vectors and the newly added `IHt,JHt` here? And why this could avoid the indirect addressing mentioned in your PR?

Reply:
So the original mPtr, mVal, and I2H indices was an ugly workaround to the fact that the H vector was stored in CSR format not the CSC format needed for the H^t operator.  In CSR format, there is a pointer to the beginning of the group of rows.  Because I needed to access the H matrix in groups of columns, I had to create some indirect address arrays that pointed to the various entries related to a group of columns.  These extra arrays were necessary in order to allow parallelization on the GPU.  There was a lower memory solution that was in the initial version of the code, but that implementation prevented any thread-based parallelism on the CPU or GPU.   While I initially did not want to explicitly store the H^t matrix, the necessary indirect address to perform the operation in parallel ended up consuming somewhat more memory than the explicit storage of H^t in CSR format.  Does that make sense?

Reply:
Thanks John for the detailed explanation. Now it is clear to me.